xla/service/gpu/transforms/async_wrapper_test.cc

```diff
@@ -81,5 +82,132 @@ TEST_F(AsyncWrapperTest, BasicFusion) {
   EXPECT_TRUE(LiteralTestUtil::Equal(expected, result));
 }
 
+TEST_F(AsyncWrapperTest, BasicFusionWithinWhile) {
+  const char* hlo_text = R"(
+  HloModule m
+
+  double1 {
+    p0 = f32[1] parameter(0)
+    ROOT add = f32[1] add(p0, p0)
+  }
+
+  double2 {
+    p0 = f32[1] parameter(0)
+    ROOT add = f32[1] add(p0, p0)
+  }
+
+  half1 {
+    p0 = f32[1] parameter(0)
+    c2 = f32[1] constant({2})
+    ROOT add = f32[1] divide(p0, c2)
+  }
+
+  body {
+    param = (f32[1], s32[]) parameter(0)
+    p0 = f32[1] get-tuple-element(param), index=0
+    agg1 = f32[1] fusion(p0), kind=kLoop, calls=double1
+    agg2 = f32[1] fusion(p0), kind=kLoop, calls=double2
+    done = f32[1] add(agg1, agg2)
+    iter = s32[] get-tuple-element(param), index=1
+    c1 = s32[] constant(1)
+    add = s32[] add(iter, c1)
+    ROOT tuple = (f32[1], s32[]) tuple(done, add)
+  }
+
+  condition {
+    param.1 = (f32[1], s32[]) parameter(0)
+    iter.1 = s32[] get-tuple-element(param.1), index=1
+    c4 = s32[] constant(4)
+    ROOT compare = pred[] compare(iter.1, c4), direction=LT
+  }
+
+  ENTRY main {
+    c0 = s32[] constant(0)
+    p0.1 = f32[1] parameter(0)
+    agg3 = f32[1] fusion(p0.1), kind=kLoop, calls=half1
+    tuple = (f32[1], s32[]) tuple(agg3, c0)
+    while = (f32[1], s32[]) while(tuple), body=body, condition=condition
+    ROOT done.1 = f32[1] get-tuple-element(while), index=0
+  })";
+
+  std::unique_ptr<VerifiedHloModule> module =
+      ParseAndReturnVerifiedModule(hlo_text).value();
+
+  AsyncWrapper wrapper(HloPredicateIsOp<HloOpcode::kFusion>);
+  EXPECT_THAT(wrapper.HloModulePass::Run(module.get()), IsOkAndHolds(true));
+  EXPECT_EQ(CountAsyncInstructions(module->entry_computation()), 2);
+  HloInstruction* while_op = hlo_query::FindInstruction(
+      module->entry_computation(), HloOpcode::kWhile);
+  ASSERT_NE(while_op, nullptr);
+  EXPECT_EQ(CountAsyncInstructions(while_op->while_body()), 4);
+
+  Literal argument = LiteralUtil::CreateR1<float>({1.0});
+  Literal expected = LiteralUtil::CreateR1<float>({128.0});
+
+  Literal result = ExecuteNoHloPasses(std::move(module), {&argument});
+  EXPECT_TRUE(LiteralTestUtil::Equal(expected, result));
```

Comment:
Do we need to run this? Seems like we could get away with checking what the resulting HLO looks like, right?

Reply:
Yes, I updated the test and removed the execution.